tests: adding system/tests/readme.rst as a quick primer (#7681)
SSSD/sssd: src/tests/system/tests/readme.rst

SSSD System Tests
=================

Tests are written in Python using our `SSSD <https://tests.sssd.io/en/latest/>`__ framework and `pytest <https://docs.pytest.org/en/stable/>`__ framework and most notably `pytest mh <https://github.com/next-actions/pytest-mh>`__  plug-in amongst others. For more information please visit our `SSSD framework wiki <https://tests.sssd.io/en/latest/>`__.

.. note::

>     We are in the progress of overhauling our all of our tests. This large housekeeping effort, there is an emphasis on providing excellent test coverage, clear concise documentation and easy to maintain test code. These tests are not just system tests but will contain some integration, but will be marked.

Reviewer: `our all of our tests`

Reviewer: Two `but` on the same (last) sentence.

Reviewer: `This large housekeeping effort, there is...`

Shouldn' t it be 'In this large housekeeping...` or `With this large...`?

Author: I have reworded the entire thing.

Reviewer: Nice! Thanks.

Should we mention how tests will be marked?

* test_services.py, contains tests specific to the SSSD and processes.
* test_responders.py, contains responders tests, NSS, PAM and others.

.. note::

    Samba Active Directory is the only AD topology used when testing upstream. Please use 'AnyAD' topology if early testing is needed.

Test Scope
**********

Tests are written to check a specific user scenario, parameter, issue or bug. Often it will be faster to merge tests into one, or when test cases overlap. For simplicity, every test case is small and covers a specific case.

Test Naming
***********

Pytest requires that tests names and file names are pre-fixed with *test_*. Test names have been standardize to then contain the file name and describe what the test code does and may not match related bugs, issues or tickets.

Docstrings
**********

Each test is required to contain the following docstring fields.

* title: Required, descriptive abbreviated test name, the name should give you a good idea of what it covers.
* description: Optional, provide more detail about the test if necessary.
* setup: Required, incremented list of steps needed before the test is executed.

> * steps: Required, incremented list of the test steps. The first step should be the the test.

Reviewer: `the the`

What do you mean by `The first step should be the test`?

Author: That is unclear, if the test is looking up a user, the step should begin there. Where as if the test is testing the startup of SSSD, the test should begin with SSSD starting.  Reworded, 

`
* setup: Required, incremented list of steps needed before the test scenario can be executed.
* steps: Required, incremented list of the test steps, beginning with the test scenario.
 `
?

Reviewer: Good enough for me.

        :setup:
            1. Create user
            2. Start SSSD
        :steps:
            1. Authenticate user with correct password
            2. Authenticate user with incorrect password
        :expectedresults:
            1. Authentication is successful
            2. Authentication is unsuccessful
        :customerscenario: False
        """
        provider.user("user1").add(password="Secret123")

        client.sssd.start(service_user=sssd_service_user)

        assert client.auth.parametrize(method).password("user1", "Secret123"), "User failed login!"
        assert not client.auth.parametrize(method).password(
            "user1", "NOTSecret123"
        ), "User logged in with an invalid password!"

.. note::

    Test code should follow the steps in order, making it easy to follow. Strip anything that is not relevant to the test, like extra configuration parameters, unused users or groups. Exceptions are fine, kindly comment the reason. Generally, language should be clear and short enough to comprehend the case but should be reduced if it becomes overly complicated with detail.

Parameterization
****************

>     Tests can be parameterized to reduce the volume of test cases. In the following example, a total of four scenarios will be executed from this single test: 'su:root, su:sssd, ssh:root, su:root'. Previously, four test cases would have been written to provide the same coverage.

Reviewer: In `su:root, su:sssd, ssh:root, su:root`, the fourth element should be `ssh:sssd`.

>     Tests can be parameterized to reduce the volume of test cases. In the following example, a total of four scenarios will be executed from this single test: 'su:root, su:sssd, ssh:root, su:sssd'. Previously, four test cases would have been written to provide the same coverage.

Reviewer: The fourth element in the list should be `ssh:sssd`, not `su:sssd`

SSSD System Tests
=================

This directory contains tests currently used in upstream and downstream continuous integration (CI) testing.

> Tests are written in Python using our `SSSD <https://tests.sssd.io/en/latest/>`__ framework and `pytest <https://docs.pytest.org/en/stable/>`__ framework and most notably `pytest multihost <https://github.com/next-actions/pytest-mh>`__  plug-in amongst others. For more information please visit our `SSSD framework wiki <https://tests.sssd.io/en/latest/>`__.

Reviewer 2: ```suggestion
Tests are written in Python using our `SSSD <https://tests.sssd.io/en/latest/>`__ framework and `pytest <https://docs.pytest.org/en/stable/>`__ framework and most notably `pytest mh <https://github.com/next-actions/pytest-mh>`__  plug-in amongst others. For more information please visit our `SSSD framework wiki <https://tests.sssd.io/en/latest/>`__.
```

.. note::

    We are in the progress of overhauling our all of our tests. This large housekeeping effort, there is an emphasis on providing excellent test coverage, clear concise documentation and easy to maintain test code. These tests are not just system tests but will contain some integration, but will be marked.

Test Categories
***************

> Tests can be written for a specific identity provider; IPA, LDAP, KRB, Samba or AD. Additionally they can also be written to use a generic provider which will execute the test against all providers. Tests are organized by user scenarios, depending on the case, they can be organized by providers or feature.

Reviewer 2: ```suggestion
Tests can be written for a specific identity provider; IPA, LDAP, KRB, Samba or AD. When possible, they should be written using a generic provider which will execute the test against all providers. Tests are organized by user scenarios, depending on the case, they can be organized by providers or feature.
```
